### PR TITLE
Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift

### PR DESCRIPTION
**Pull request by soumith**
This PR implements Batch Normalization as described in the paper:
   "Batch Normalization: Accelerating Deep Network Training
                         by Reducing Internal Covariate Shift"
                   by Sergey Ioffe, Christian Szegedy

   For Convolution layers, see SpatialBatchNormalization.lua
   For Fully-connected layers, see BatchNormalization.lua
   The operation implemented is:
```
   y =     ( x - mean(x) )
           ------------------------   * gamma + beta
          standard-deviation(x)
```
   where gamma and beta are learnable parameters.
   The learning of gamma and beta is optional.
   Usage:
```
 Fully connected:
   with    learnable parameters:   nn.BatchNormalization(N [, eps]) where N = dimensionality of input
   without learnable parameters: nn.BatchNormalization(0 [, eps])
   eps is a small value added to the standard-deviation to avoid divide-by-zero.
       Defaults to 1e-5

Convolution:
   with    learnable parameters:   nn.SpatialBatchNormalization(N [, eps]) where N = number of feature maps
   without learnable parameters: nn.SpatialBatchNormalization(0 [, eps])
   eps is a small value added to the standard-deviation to avoid divide-by-zero.
       Defaults to 1e-5

```

### CODE REVIEW

File: doc/convolution.md [hunk 1]
```diff
@@ -500,6 +501,43 @@ w2=image.display(processed)
...
+<a name="nn.SpatialBatchNormalization"/>
+## SpatialBatchNormalization ##
+
+`module` = `nn.SpatialBatchNormalization(N [,eps])`
+ where N = number of input feature maps
+giving N = 0 disables the learnable affine transform.
+eps is a small value added to the standard-deviation to avoid divide-by-zero. Defaults to 1e-5
+
+Implements Batch Normalization as described in the paper:
+   "Batch Normalization: Accelerating Deep Network Training
```
Comment (nicholas-leonard) [inline doc/convolution.md]: Add hyperlink?

**Comment by nicholas-leonard** [comment]
Have been reviewing it for the past hour or so. Unit tests pass. The code looks great.

**Comment by szagoruyko** [comment]
great! what about inference mode, how to process multiple mini-batches to get average E[x] and Var[x]?
also any plans to update imagenet example with this?

**Comment by soumith** [comment]
no inference mode yet. i will probably add it eventually if no one else does. this is just for everyone to play with have have fun. 
https://github.com/soumith/imagenet-multiGPU.torch/blob/master/models/alexnetowtbn_cudnn.lua

the batch-norm version. training error goes down fast but overfits, feel free to play around with it. I am doing a few other things for the next week or two, so wont be able to add more concrete examples of using batchnorm to the imagenet example.

**Comment by soumith** [comment]
Added inference mode (not exactly to the paper but works great in practice!)

**Comment by jonathantompson** [comment]
Hey @soumith.

1 quick question (not comments about your code, just questions about the paper):

The authors ignore correlation between activations.  Have you tried using this in networks whose activations are known to be strongly correlated?  Or just imagenet classification networks?

**Comment by soumith** [comment]
I've only tried it on image classification problems and it works like magic sauce. Durk kingma also commented somewhere that it is awesome for variational AEs and anywhere where you might also see gradient explosions.

**Comment by jonathantompson** [comment]
Nice.  OK I will free up some Titan's an try it :-)

**Comment by zhongwen** [comment]
Hi @soumith , it might be better to replace running_std with running_var
